ethereum/core/src/main/java/org/hyperledger/besu/ethereum/forkid/ForkIdManager.java

```diff
@@ -34,52 +36,87 @@
+  private final List<ForkId> blockNumbersForkIds;
+  private final List<ForkId> timestampsForkIds;
 
-  private final List<Long> forkBlockNumbers;
-  private final LongSupplier chainHeadSupplier;
+  private final List<Long> blockNumberForks;
+  private final List<Long> timestampForks;
+
+  private final Supplier<BlockHeader> chainHeadSupplier;
   private final long forkNext;
   private final boolean onlyZerosForkBlocks;
   private final long highestKnownFork;
   private Bytes genesisHashCrc;
   private final boolean legacyEth64;
 
   public ForkIdManager(
-      final Blockchain blockchain, final List<Long> nonFilteredForks, final boolean legacyEth64) {
+      final Blockchain blockchain,
+      final List<Long> blockNumberForks,
+      final List<Long> timestampForks,
+      final boolean legacyEth64) {
     checkNotNull(blockchain);
-    checkNotNull(nonFilteredForks);
-    this.chainHeadSupplier = blockchain::getChainHeadBlockNumber;
+    checkNotNull(blockNumberForks);
+    this.chainHeadSupplier = blockchain::getChainHeadHeader;
     this.genesisHash = blockchain.getGenesisBlock().getHash();
-    this.forkIds = new ArrayList<>();
+    this.blockNumbersForkIds = new ArrayList<>();
+    this.timestampsForkIds = new ArrayList<>();
     this.legacyEth64 = legacyEth64;
-    this.forkBlockNumbers =
-        nonFilteredForks.stream()
+    this.blockNumberForks =
+        blockNumberForks.stream()
             .filter(fork -> fork > 0L)
             .distinct()
             .sorted()
             .collect(Collectors.toUnmodifiableList());
-    this.onlyZerosForkBlocks = nonFilteredForks.stream().allMatch(value -> 0L == value);
+    this.timestampForks =
+        timestampForks.stream()
+            .filter(fork -> fork > 0L)
+            .distinct()
+            .sorted()
+            .collect(Collectors.toUnmodifiableList());
+    this.onlyZerosForkBlocks =
+        Stream.concat(blockNumberForks.stream(), timestampForks.stream())
+            .allMatch(value -> 0L == value);
     this.forkNext = createForkIds();
+    final long highestKnownBlockFork =
+        !blockNumberForks.isEmpty() ? blockNumberForks.get(blockNumberForks.size() - 1) : 0L;
     this.highestKnownFork =
-        !forkBlockNumbers.isEmpty() ? forkBlockNumbers.get(forkBlockNumbers.size() - 1) : 0L;
+        !timestampForks.isEmpty()
+            ? timestampForks.get(timestampForks.size() - 1)
+            : highestKnownBlockFork;
   }
 
   public ForkId getForkIdForChainHead() {
     if (legacyEth64) {
-      return forkIds.isEmpty() ? null : forkIds.get(forkIds.size() - 1);
+      return blockNumbersForkIds.isEmpty()
+          ? null
+          : blockNumbersForkIds.get(blockNumbersForkIds.size() - 1);
     }
-    final long head = chainHeadSupplier.getAsLong();
-    for (final ForkId forkId : forkIds) {
-      if (head < forkId.getNext()) {
+    final BlockHeader header = chainHeadSupplier.get();
+    for (final ForkId forkId : blockNumbersForkIds) {
```

Comment:
this could/should be pulled from all forks if we allow timestamps =1 (or 0).  Since ethereum started well after epoch time.  It is a corner case, but test networks might make some interesting assumptions for their genesis